daemon/daemon.go

```diff
@@ -694,98 +689,13 @@ func (d *Daemon) createNodeConfigHeaderfile() error {
-	}
-
-	fw.WriteString(common.FmtDefineComma("HOST_IP", hostIP))
-	fmt.Fprintf(fw, "#define HOST_ID %d\n", identity.GetReservedID(labels.IDNameHost))
-	fmt.Fprintf(fw, "#define WORLD_ID %d\n", identity.GetReservedID(labels.IDNameWorld))
-	fmt.Fprintf(fw, "#define HEALTH_ID %d\n", identity.GetReservedID(labels.IDNameHealth))
-	fmt.Fprintf(fw, "#define UNMANAGED_ID %d\n", identity.GetReservedID(labels.IDNameUnmanaged))
-	fmt.Fprintf(fw, "#define INIT_ID %d\n", identity.GetReservedID(labels.IDNameInit))
-	fmt.Fprintf(fw, "#define LB_RR_MAX_SEQ %d\n", lbmap.MaxSeq)
-	fmt.Fprintf(fw, "#define CILIUM_LB_MAP_MAX_ENTRIES %d\n", lbmap.MaxEntries)
-	fmt.Fprintf(fw, "#define TUNNEL_MAP %s\n", tunnel.MapName)
-	fmt.Fprintf(fw, "#define TUNNEL_ENDPOINT_MAP_SIZE %d\n", tunnel.MaxEntries)
-	fmt.Fprintf(fw, "#define PROXY_MAP_SIZE %d\n", proxymap.MaxEntries)
-	fmt.Fprintf(fw, "#define ENDPOINTS_MAP %s\n", lxcmap.MapName)
-	fmt.Fprintf(fw, "#define ENDPOINTS_MAP_SIZE %d\n", lxcmap.MaxEntries)
-	fmt.Fprintf(fw, "#define METRICS_MAP %s\n", metricsmap.MapName)
-	fmt.Fprintf(fw, "#define METRICS_MAP_SIZE %d\n", metricsmap.MaxEntries)
-	fmt.Fprintf(fw, "#define POLICY_MAP_SIZE %d\n", policymap.MaxEntries)
-	fmt.Fprintf(fw, "#define IPCACHE_MAP %s\n", ipcachemap.Name)
-	fmt.Fprintf(fw, "#define IPCACHE_MAP_SIZE %d\n", ipcachemap.MaxEntries)
-	fmt.Fprintf(fw, "#define POLICY_PROG_MAP_SIZE %d\n", policymap.ProgArrayMaxEntries)
-	fmt.Fprintf(fw, "#define SOCKOPS_MAP_SIZE %d\n", sockmap.MaxEntries)
-
-	if option.Config.DatapathMode == option.DatapathModeIpvlan {
-		fmt.Fprintf(fw, "#define ENABLE_SECCTX_FROM_IPCACHE 1\n")
-	}
-
-	if option.Config.PreAllocateMaps {
-		fmt.Fprintf(fw, "#define PREALLOCATE_MAPS 1\n")
-	}
-
-	fmt.Fprintf(fw, "#define EVENTS_MAP %s\n", "cilium_events")
-	fmt.Fprintf(fw, "#define POLICY_CALL_MAP %s\n", policymap.CallMapName)
-	fmt.Fprintf(fw, "#define PROXY4_MAP cilium_proxy4\n")
-	fmt.Fprintf(fw, "#define PROXY6_MAP cilium_proxy6\n")
-	fmt.Fprintf(fw, "#define EP_POLICY_MAP %s\n", eppolicymap.MapName)
-	fmt.Fprintf(fw, "#define LB6_REVERSE_NAT_MAP cilium_lb6_reverse_nat\n")
-	fmt.Fprintf(fw, "#define LB6_SERVICES_MAP cilium_lb6_services\n")
-	fmt.Fprintf(fw, "#define LB6_RR_SEQ_MAP cilium_lb6_rr_seq\n")
-	fmt.Fprintf(fw, "#define LB4_REVERSE_NAT_MAP cilium_lb4_reverse_nat\n")
-	fmt.Fprintf(fw, "#define LB4_SERVICES_MAP cilium_lb4_services\n")
-	fmt.Fprintf(fw, "#define LB4_RR_SEQ_MAP cilium_lb4_rr_seq\n")
-
-	fmt.Fprintf(fw, "#define TRACE_PAYLOAD_LEN %dULL\n", option.Config.TracePayloadlen)
-	fmt.Fprintf(fw, "#define MTU %d\n", d.mtuConfig.GetDeviceMTU())
-
-	if option.Config.EnableIPv4 {
-		fmt.Fprintf(fw, "#define ENABLE_IPV4 1\n")
-	}
-
-	if option.Config.EnableIPv6 {
-		fmt.Fprintf(fw, "#define ENABLE_IPV6 1\n")
 	}
+	defer f.Close()
 
-	if option.Config.EnableIPSec {
-		fmt.Fprintf(fw, "#define ENABLE_IPSEC 1\n")
+	if err = d.datapath.WriteNodeConfig(f, &d.nodeDiscovery.localConfig); err != nil {
```

Comment:
Is there any chance that `d.nodeDiscovery` could be nil? We would panic in that case. Maybe check if it's set to nil and if so return an error?

Reply:
This should never happen, because it is always initialized when we create the daemon object in `NewDaemon()` (and `localConfig` similarly is initialized in `newNodeDiscovery()`). Ideally we should never get into such situations if we follow the RAII principle (Resource Acquisition Is Initialization), ie "when you create the object, initialize all of the fields to reasonable values". We appear to be doing so in this case. Glad you asked though, was not paying close attention to this.